test/fake/client.go

```diff
@@ -0,0 +1,229 @@
+	"k8sgateway/internal/token"
+)
+
+// Client simulates a Twingate Client, authenticating and forwarding kubectl requests to the Gateway.
+//
+// Client is implemented as a TCP proxy. On receiving a connection, it opens a connection to the
+// Gateway, authenticates using a CONNECT request, and then forwards TCP data.
+//
+// kubectl CLI needs to connect to this Client's listener address.
+type Client struct {
+	Listener net.Listener
+	URL      string
+
+	user          *token.User
+	proxyAddress  string
+	controllerURL string
+	apiServerURL  string
+
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+func NewClient(user *token.User, proxyAddress, controllerURL, apiServerURL string) *Client {
+	logger := zap.Must(zap.NewDevelopment()).Named(fmt.Sprintf("client-%s-%s", user.ID, user.Username))
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		logger.Fatal("Failed to listen", zap.Error(err))
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &Client{
+		Listener:      listener,
+		URL:           "https://" + listener.Addr().String(),
+		proxyAddress:  proxyAddress,
+		controllerURL: controllerURL,
+		apiServerURL:  apiServerURL,
+		user:          user,
+		cancel:        cancel,
+		wg:            &sync.WaitGroup{},
+		logger:        logger,
+	}
+	go c.serve(ctx)
+
+	return c
+}
+
+// Close gracefully shuts down the client.
+//
+// It will close the listener and wait for all existing connections to complete. It does not
+// terminate existing connections forcibly i.e. it might hang indefinitely if the downstream
+// connection is not properly closed.
+func (c *Client) Close() {
+	c.cancel()
+	c.Listener.Close()
+	c.wg.Wait()
+}
+
+func (c *Client) serve(ctx context.Context) {
+	gat := c.fetchGAT()
+	if gat == "" {
+		c.logger.Error("Failed to fetch GAT")
+
+		return
+	}
+
+	for {
+		clientConn, err := c.Listener.Accept()
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				c.logger.Error("Failed to accept connection", zap.Error(err))
+			}
+
+			continue
+		}
+
+		c.wg.Add(1)
+		go c.handleConnection(ctx, clientConn, gat)
+	}
+}
+
+func (c *Client) handleConnection(ctx context.Context, clientConn net.Conn, gat string) {
+	defer clientConn.Close()
+	defer c.wg.Done()
+
+	// Proxy certs
+	caCert, _ := os.ReadFile("../data/proxy/tls.crt")
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS13,
+		ServerName: "127.0.0.1",
+		RootCAs:    caCertPool,
+	}
+
+	// Manually create a TCP connection
+	conn, err := net.Dial("tcp", c.proxyAddress)
+	if err != nil {
+		c.logger.Error("Failed to connect to proxy", zap.Error(err))
+	}
+	defer conn.Close()
+
+	// Enable outer TLS (between Twingate client and proxy)
+	proxyTLSConn := tls.Client(conn, tlsConfig)
+	if err := proxyTLSConn.Handshake(); err != nil {
+		c.logger.Error("TLS handshake failed(proxy)", zap.Error(err))
+	}
+
```

Comment:
should also add `defer proxyTLSConn.close()` here to ensure clean TLS shutdown with `close-notify`